guide/src/dev/performance.md

```diff
@@ -0,0 +1,206 @@
+2. Loading 16 bytes at a time into a SIMD register
+3. Performing parallel lookups to identify special characters
+4. Generating a bitmask indicating which bytes matched
+
+```rust
+// Example from simd.rs showing the core scanning logic
+#[target_feature(enable = "ssse3")]
+unsafe fn compute_mask(lut: &[u8; 16], bytes: &[u8], ix: usize) -> i32 {
+    let bitmap = _mm_loadu_si128(lut.as_ptr() as *const __m128i);
+    let input = _mm_loadu_si128(bytes.as_ptr().add(ix) as *const __m128i);
+    let bitset = _mm_shuffle_epi8(bitmap, input);
+    let higher_nibbles = _mm_and_si128(_mm_srli_epi16(input, 4), _mm_set1_epi8(0x0f));
+    let bitmask = _mm_shuffle_epi8(bitmask_lookup, higher_nibbles);
+    let tmp = _mm_and_si128(bitset, bitmask);
+    let result = _mm_cmpeq_epi8(tmp, bitmask);
+    _mm_movemask_epi8(result)
+}
+```
+
+This SIMD optimization can provide significant speedups when processing large documents, since character scanning is such a common operation. The code falls back to scalar processing when SIMD is not available.
+
+## Memory-Efficient String Storage
+
+The library uses a custom string type `CowStr` that can represent strings in three ways:
+
+1. `Borrowed` - Zero-copy reference to input text
+2. `Inlined` - Small strings stored directly in the enum
+3. `Boxed` - Heap allocated for larger strings
+
+```rust
+pub enum CowStr<'a> {
+    Boxed(Box<str>),
+    Borrowed(&'a str),
+    Inlined(InlineStr),
+}
+```
+
+This allows efficient handling of strings with different sizes and lifetimes:
+
+- Small strings can avoid allocation by using the inline storage
+- References to input text avoid copying when possible
+- Heap allocation is only used when necessary for large strings
+
+The `InlineStr` type can store strings up to 22 bytes (on 64-bit systems) directly in the enum variant, avoiding heap allocation for common short string cases. Refer to the [string handling](./string-handling.md) documentation for more details.
+
+## Tree Structure Optimization
+
+The AST (Abstract Syntax Tree) is stored in a vec-based tree structure that provides:
+
+1. Fast node creation during parsing
+2. Efficient tree traversal
+3. Memory locality from vector storage
+
+```rust
+pub(crate) struct Tree<T> {
+    nodes: Vec<Node<T>>,
+    spine: Vec<TreeIndex>,
+    cur: Option<TreeIndex>, 
+}
+
+pub(crate) struct Node<T> {
+    pub child: Option<TreeIndex>,
+    pub next: Option<TreeIndex>,
+    pub item: T,
+}
+```
+
+Key optimizations in the tree structure include:
+
+- Using indices instead of pointers for node references
+- Maintaining a "spine" for fast access to ancestor nodes
+- Storing nodes contiguously in a vector for better cache usage
+- Using non-zero indices to save space in option types
+
+## Single-Pass Parsing
+
+The parser uses a single pass over the input text for both block and inline parsing, avoiding the need to rescan text multiple times. This is achieved through:
```

Comment:
Untrue - there are multiple passes (two principle ones but also many smaller ones depending on how you look at it)